src/neo4j_graphrag/retrievers/cypher.py

```diff
@@ -0,0 +1,379 @@
+    CypherParameterType,
+    CypherRetrieverModel,
+    CypherSearchModel,
+    Neo4jDriverModel,
+    RawSearchResult,
+    RetrieverResultItem,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class CypherRetriever(Retriever):
+    """
+    Allows for the retrieval of records from a Neo4j database using a parameterized Cypher query.
+
+    This retriever enables direct execution of predefined Cypher queries with dynamic parameters.
+    It ensures type safety through parameter validation and provides the standard retriever result format.
+
+    Example:
+
+    .. code-block:: python
+
+        import neo4j
+        from neo4j_graphrag.retrievers import CypherRetriever
+
+        driver = neo4j.GraphDatabase.driver(URI, auth=AUTH)
+
+        # Create a retriever for finding movies by title
+        retriever = CypherRetriever(
+            driver=driver,
+            query="MATCH (m:Movie {title: $movie_title}) RETURN m",
+            parameters={
+                "movie_title": {
+                    "type": "string",
+                    "description": "Title of a movie"
+                }
+            }
+        )
+
+        # Use the retriever with specific parameter values
+        results = retriever.search(parameters={"movie_title": "The Matrix"})
+
+    Args:
+        driver (neo4j.Driver): The Neo4j Python driver.
+        query (str): Cypher query with parameter placeholders.
+        parameters (Dict[str, Dict]): Parameter definitions with types and descriptions.
+            Each parameter should have a 'type' and 'description' field.
+            Supported types: 'string', 'number', 'integer', 'boolean', 'array'.
+        result_formatter (Optional[Callable[[neo4j.Record], RetrieverResultItem]]):
+            Custom function to transform a neo4j.Record to a RetrieverResultItem.
+        neo4j_database (Optional[str]): The name of the Neo4j database to use.
+
+    Raises:
+        RetrieverInitializationError: If validation of the input arguments fail.
+    """
+
+    def __init__(
+        self,
+        driver: neo4j.Driver,
+        query: str,
+        parameters: Dict[str, Dict[str, Any]],
+        result_formatter: Optional[
+            Callable[[neo4j.Record], RetrieverResultItem]
+        ] = None,
+        neo4j_database: Optional[str] = None,
+    ) -> None:
+        # Convert parameter dictionaries to CypherParameterDefinition objects
+        param_definitions = {}
+        for param_name, param_def in parameters.items():
+            param_type = param_def.get("type", "string")
+            description = param_def.get("description", "")
+            required = param_def.get("required", True)
+
+            try:
+                param_definitions[param_name] = CypherParameterDefinition(
+                    type=param_type, description=description, required=required
+                )
+            except ValidationError as e:
+                raise RetrieverInitializationError(
+                    [
```

Comment:
I've never used this syntax but I think `from e` will show the same informations (loc, etc)? In such case we do not need `ErrorDestail`, which looks like an internal pydantic thing.